arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,234 @@
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBChecksumEnv.h"
+
+#include "Basics/Exceptions.h"
+#include "Basics/files.h"
+#include "Basics/FileUtils.h"
+#include "Basics/MutexLocker.h"
+#include "Logger/LogMacros.h"
+
+namespace arangodb::checksum {
+
+ChecksumCalculator::ChecksumCalculator()
+    :
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+      _context(EVP_MD_CTX_new()) {
+#else
+      _context(EVP_MD_CTX_create()) {
+#endif
+  if (_context == nullptr) {
+    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
+  }
+  if (EVP_DigestInit_ex(_context, EVP_sha256(), nullptr) == 0) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL,
+                                   "unable to initialize SHA256 processor");
+  }
+}
+
+ChecksumCalculator::~ChecksumCalculator() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void ChecksumCalculator::update(char const* buffer, size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  EVP_DigestUpdate(_context, static_cast<void const*>(buffer), n);
+}
+
+bool ChecksumHelper::isFileNameSst(std::string const& fileName) noexcept {
+  return TRI_Basename(fileName).size() > 4 &&
+         (fileName.compare(fileName.size() - 4, 4, ".sst") == 0);
+}
+
+bool ChecksumHelper::writeShaFile(std::string const& fileName,
+                                  std::string const& checksum) {
+  TRI_ASSERT(isFileNameSst(fileName));
+
+  std::string shaFileName = fileName.substr(0, fileName.size() - 4);
+  shaFileName += ".sha.";
+  shaFileName += checksum;
+  shaFileName += ".hash";
+  LOG_TOPIC("80257", DEBUG, arangodb::Logger::ENGINES)
+      << "shaCalcFile: done " << fileName << " result: " << shaFileName;
+  auto res = TRI_WriteFile(shaFileName.c_str(), "", 0);
+  if (res == TRI_ERROR_NO_ERROR) {
+    MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+    _sstFileNamesToHashes.try_emplace(TRI_Basename(fileName), checksum);
+    return true;
+  }
+
+  LOG_TOPIC("8f7ef", WARN, arangodb::Logger::ENGINES)
+      << "shaCalcFile: TRI_WriteFile failed with " << res << " for "
+      << shaFileName;
+  return false;
+}
+
+std::string ChecksumCalculator::computeChecksum() {
```

Comment:
Can the methods in this file be ordered by class? It is hard to get an overview as it is right now, because methods of different classes are interleaved in the file.